File: core-banking-prototype-laravel-main/core-banking-prototype-laravel-main/sdks/php/src/Resources/Webhooks.php
<?php

namespace FinAegis\Resources;

use FinAegis\Models\PaginatedResponse;
use FinAegis\Models\Webhook;

class Webhooks extends BaseResource
{
    /**
     * List all webhooks.
     *
     * @param  int  $page  Page number
     * @param  int  $perPage  Items per page
     */
    public function list(int $page = 1, int $perPage = 20): PaginatedResponse
    {
        $response = $this->get('/webhooks', ['page' => $page, 'per_page' => $perPage]);

        return new PaginatedResponse($response, Webhook::class);
    }

    /**
     * Create a new webhook.
     *
     * @param  string  $name  Webhook name
     * @param  string  $url  Webhook URL
     * @param  array  $events  List of events to subscribe to
     * @param  array|null  $headers  Optional custom headers
     * @param  string|null  $secret  Optional secret for signature verification
     */
    public function create(
        string $name,
        string $url,
        array $events,
        ?array $headers = null,
        ?string $secret = null
    ): Webhook {
        $data = [
            'name' => $name,
            'url' => $url,
            'events' => $events,
        ];

        if ($headers !== null) {
            $data['headers'] = $headers;
        }

        if ($secret !== null) {
            $data['secret'] = $secret;
        }

        $response = $this->post('/webhooks', $data);

        return new Webhook($response['data']);
    }

    /**
     * Get webhook details.
     *
     * @param  string  $webhookId  Webhook UUID
     */
    public function get(string $webhookId): Webhook
    {
        $response = $this->get("/webhooks/{$webhookId}");

        return new Webhook($response['data']);
    }

    /**
     * Update a webhook.
     *
     * @param  string  $webhookId  Webhook UUID
     * @param  array  $data  Update data
     */
    public function update(string $webhookId, array $data): Webhook
    {
        $response = $this->put("/webhooks/{$webhookId}", $data);

        return new Webhook($response['data']);
    }

    /**
     * Delete a webhook.
     *
     * @param  string  $webhookId  Webhook UUID
     */
    public function delete(string $webhookId): array
    {
        return $this->delete("/webhooks/{$webhookId}");
    }

    /**
     * Get webhook delivery history.
     *
     * @param  string  $webhookId  Webhook UUID
     * @param  int  $page  Page number
     * @param  int  $perPage  Items per page
     */
    public function getDeliveries(string $webhookId, int $page = 1, int $perPage = 20): array
    {
        return $this->get("/webhooks/{$webhookId}/deliveries", ['page' => $page, 'per_page' => $perPage]);
    }

    /**
     * Get available webhook events.
     */
    public function getEvents(): array
    {
        $response = $this->get('/webhooks/events');

        return $response['data'];
    }

    /**
     * Verify webhook signature.
     *
     * @param  string  $payload  Webhook payload
     * @param  string  $signature  Webhook signature
     * @param  string  $secret  Webhook secret
     */
    public static function verifySignature(string $payload, string $signature, string $secret): bool
    {
        $expectedSignature = hash_hmac('sha256', $payload, $secret);

        return hash_equals($expectedSignature, $signature);
    }
}

File: core-banking-prototype-laravel-main/core-banking-prototype-laravel-main/sdks/python/finaegis/resources/webhooks.py
"""
Webhooks resource for the FinAegis SDK
"""

from typing import List, Optional, Dict, Any
from ..types import Webhook, PaginatedResponse
from .base import BaseResource


class WebhooksResource(BaseResource):
    """Manage webhooks in the FinAegis platform."""
    
    def list(self, page: int = 1, per_page: int = 20) -> PaginatedResponse:
        """
        List all webhooks.
        
        Args:
            page: Page number
            per_page: Items per page
            
        Returns:
            PaginatedResponse containing Webhook objects
        """
        response = self._get('/webhooks', params={'page': page, 'per_page': per_page})
        return PaginatedResponse.from_dict(response, Webhook)
    
    def create(
        self,
        name: str,
        url: str,
        events: List[str],
        headers: Optional[Dict[str, str]] = None,
        secret: Optional[str] = None
    ) -> Webhook:
        """
        Create a new webhook.
        
        Args:
            name: Webhook name
            url: Webhook URL
            events: List of events to subscribe to
            headers: Optional custom headers
            secret: Optional secret for signature verification
            
        Returns:
            Created Webhook object
        """
        data = {
            'name': name,
            'url': url,
            'events': events
        }
        if headers:
            data['headers'] = headers
        if secret:
            data['secret'] = secret
            
        response = self._post('/webhooks', data)
        return Webhook.from_dict(response['data'])
    
    def get(self, webhook_id: str) -> Webhook:
        """
        Get webhook details.
        
        Args:
            webhook_id: Webhook ID
            
        Returns:
            Webhook object
        """
        response = self._get(f'/webhooks/{webhook_id}')
        return Webhook.from_dict(response['data'])
    
    def update(
        self,
        webhook_id: str,
        name: Optional[str] = None,
        url: Optional[str] = None,
        events: Optional[List[str]] = None,
        headers: Optional[Dict[str, str]] = None,
        is_active: Optional[bool] = None
    ) -> Webhook:
        """
        Update a webhook.
        
        Args:
            webhook_id: Webhook ID
            name: New webhook name
            url: New webhook URL
            events: New list of events
            headers: New custom headers
            is_active: Whether the webhook is active
            
        Returns:
            Updated Webhook object
        """
        data: Dict[str, Any] = {}
        if name is not None:
            data['name'] = name
        if url is not None:
            data['url'] = url
        if events is not None:
            data['events'] = events
        if headers is not None:
            data['headers'] = headers
        if is_active is not None:
            data['is_active'] = is_active
            
        response = self._put(f'/webhooks/{webhook_id}', data)
        return Webhook.from_dict(response['data'])
    
    def delete(self, webhook_id: str) -> Dict[str, str]:
        """
        Delete a webhook.
        
        Args:
            webhook_id: Webhook ID
            
        Returns:
            Success message
        """
        return self._delete(f'/webhooks/{webhook_id}')
    
    def get_deliveries(self, webhook_id: str, page: int = 1, per_page: int = 20) -> Dict[str, Any]:
        """
        Get webhook delivery history.
        
        Args:
            webhook_id: Webhook ID
            page: Page number
            per_page: Items per page
            
        Returns:
            Paginated delivery history
        """
        response = self._get(
            f'/webhooks/{webhook_id}/deliveries',
            params={'page': page, 'per_page': per_page}
        )
        return response
    
    def get_events(self) -> Dict[str, List[Dict[str, str]]]:
        """
        Get available webhook events.
        
        Returns:
            Dictionary of events grouped by category
        """
        response = self._get('/webhooks/events')
        return response['data']